presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java

```diff
@@ -44,10 +45,25 @@ private Util()
      * <p>
      * If all inputs are used, return Optional.empty() to indicate that no pruning is necessary.
      */
-    public static Optional<Set<VariableReferenceExpression>> pruneInputs(Collection<VariableReferenceExpression> availableInputs, Collection<Expression> expressions, TypeProvider types)
+    public static Optional<Set<VariableReferenceExpression>> pruneInputs(
+            Collection<VariableReferenceExpression> availableInputs,
+            Collection<RowExpression> expressions,
+            TypeProvider types)
     {
         Set<VariableReferenceExpression> availableInputsSet = ImmutableSet.copyOf(availableInputs);
-        Set<VariableReferenceExpression> prunedInputs = Sets.filter(availableInputsSet, SymbolsExtractor.extractUniqueVariable(expressions, types)::contains);
+        Set<VariableReferenceExpression> referencedInputs;
+        if (expressions.stream().anyMatch(OriginalExpressionUtils::isExpression)) {
+            // TODO remove once all pruneInputs rules are below translateExpressions.
+            // We should either contain all original expression or all translated row expression.
```

Comment:
Do a check state here

Reply:
We cannot, `pruneInputs` is used multiple times in `planOptimizers`. Unless we move `translateExpressions` way up to the beginning, we have to handle both `OriginalExpression` and `RowExpression`.

Reply:
You have stated "We should either contain all original expression or all translated row expression." So I guess we could do a check here.

Reply:
`castToExpression` already did the check. If any of expression is not `OriginalExpression`, `castToExpression` will fail.